src/magic_folder/scanner.py

```diff
@@ -108,10 +110,13 @@ def process(path):
             d.addErrback(write_failure)
             results.append(d)
 
+        # XXX update/use IStatus to report scan start/end
+
         with start_action(action_type="scanner:find-updates"):
             yield find_updated_files(self._cooperator, self._config, process)
             yield gatherResults(results)
-        # XXX update/use IStatus to report scan start/end
+        # if we aren't already doing uploads, start some now
+        self._uploader_service.perform_upload()
```

Comment:
i think that's too granular, and we should file an enhancement ticket to have a "do uploads now" HTTP API instead -- if you're already wanting to use the low-level API of creating snapshots by hand then you can also declare when they should start (by calling the "do uploads" API). Anyway AFAIK gridsync isn't using the "create a snapshot" API.

Reply:
If we are trying to get uploads started sooner, we should start uploading a file as soon as we've taken a snapshot of it (and before we finished taking snapshots of all the other files in the folder that may need snapshots). That requires triggering from when we add a snapshot rather than here.

I agree that that will cause too many requests to upload (so it would be good to de-duplicate them), but I think living with the extra processing loops is worth it, to upload things ASAP. Long-term we'd probably want these triggered-by-snapshot upload pass to only look at *new* local snapshots, and not existing ones that failed to upload (leaving those to the periodic scan).

Reply:
...but you also don't like the de-duplication stuff that already existed here, and demanded a lock-based implementation instead.

Reply:
Well, the existing de-duplication code doesn't handle ensuring there is *at least* once additional call to `.upload_local_snapshots` after the currently running one completes. That is necessary for correctness of triggering an upload immediately. Using a lock does give at least that, if not the at-most-one; that is, a lock is correct but inefficient while the existing code was incorrect.

I'd prefer for the de-duplication to be implemented as a new primitive anyway, rather than open-coded, so using a lock as an intermediate step seems reasonable.

In any case, I think the call-asap, and ensure-we-call-upload-again are the important things to have.

Reply:
Apologies if I wasn't clear, but I think the relevant code should be in `LocalSnapshotService` or `LocalSnapshotCreator` not the scanner. The scanner shouldn't need to know how to coordinate `LocalSnapshotService` and `UploaderService`. I think this PR shouldn't involve changing the scanner code at all.

Reply:
why shouldn't it co-ordinate? 
(i know you like the idea of the low-level add-snapshot API also starting uploads, but ...)

Reply:
okay, practically-speaking this is a whole bunch more test-code, and messes up the tests that create local snapshots and then analyze them (because they get uploaded before those tests can do anything).

So, I think the explicit low-level API (if any) is better, at least immediately.  If there's a strong need to have the "create-snapshot" API also do uploads, that should be its own PR. I tried to reflect this in a comment.

Reply:
> why shouldn't it co-ordinate?

Separation of concerns.

> (i know you like the idea of the low-level add-snapshot API also starting uploads, but ...)

That has very little weight on why I think this is the wrong place to be triggering an upload pass.

> okay, practically-speaking this is a whole bunch more test-code, and messes up the tests that create local snapshots and then analyze them (because they get uploaded before those tests can do anything).

Yes, it requires some test changes, but the scanner version does too. If you use the same approach you used for the scanner, it wouldn't mess up the local snapshot tests, as the call would not actually do anything.

Reply:
> Separation of concerns.

Whether we add "uploader" to the local-snapshot-creator or scanner is pretty arbitrary and one could use that justification for either one.

Reply:
> If you use the same approach you used for the scanner, it wouldn't mess up the local snapshot tests, as the call would not actually do anything.

Not exactly because of the change of semantics to the Web API. Which I disagree with anyway. And the only (existing) test knob at that level is "all services disabled, or none" -- not "turn off the uploader service, only". Nobody has asked for "create a snapshot" to also cause uploads, so that's pretty out-of-scope for this regardless.

So, leaving this in the scanner for now and if we find its somehow better in the other place, we can followup (and one could  start with https://github.com/meejah/magic-folder/tree/513.upload-even-sooner for beginning).

Reply:
> pretty out-of-scope for this regardless

I don't think picking a different implementation strategy is something "out-of-scope".